src/bootstrap/src/ferrocene/code_coverage.rs

```diff
@@ -1,90 +1,73 @@
+    pub(crate) fn new(
+        builder: &'a Builder<'a>,
+        cargo: &mut Cargo,
+        target: TargetSelection,
+        name: &str,
+    ) -> Self {
+        builder.ensure(Llvm { target });
+        let llvm_bin_dir = builder.llvm_out(target).join("bin");
+
+        let profraw_dir = builder.tempdir().join(format!("ferrocene-profraw-{name}"));
+        let profdata_file = builder
+            .out
+            .join(target.triple)
+            .join("ferrocene")
+            .join("coverage")
+            .join(format!("{name}.profdata"));
+
+        if profraw_dir.exists() {
+            builder.remove_dir(&profraw_dir);
+        }
+        if profdata_file.exists() {
+            builder.remove(&profdata_file);
+        }
 
-        cargo.arg("--target-dir");
-        cargo.arg(&*target_dir.to_string_lossy());
-        cargo.arg("--no-default-features");
+        builder.create_dir(&profraw_dir);
+        builder.create_dir(profdata_file.parent().unwrap());
 
-        BootstrapCommand::from(cargo).fail_fast().run(builder);
+        let profraw_file_template = profraw_dir.join("%m_%p.profraw");
+        cargo.env("LLVM_PROFILE_FILE", profraw_file_template);
 
-        let cargo_dir = if builder.config.rust_optimize.is_release() { "release" } else { "debug" };
+        instrument_coverage(cargo);
 
-        let lib_dir = target_dir.join(&*target.triple).join(cargo_dir);
-        let lib_path = lib_dir.join("libprofiler_builtins.rlib");
-        let profiler_builtins_no_core_path = lib_dir.join("libprofiler_builtins_no_core.rlib");
+        Self { builder, llvm_bin_dir, profraw_dir, profdata_file }
+    }
 
-        if !builder.config.dry_run() {
-            std::fs::rename(&lib_path, &profiler_builtins_no_core_path)
-                .expect("Could not rename the profiler_builtins_no_core library");
-            assert!(
-                profiler_builtins_no_core_path.exists() && profiler_builtins_no_core_path.is_file()
-            );
+    pub(crate) fn post_process(self) {
+        if self.builder.config.dry_run() {
+            return;
         }
-        let mut instrument_coverage_flags = InstrumentationCoverageFlags { rustflags: Vec::new() };
 
-        instrument_coverage_flags.rustflags.push("-Cinstrument-coverage".into());
-        instrument_coverage_flags.rustflags.push("--extern".into());
-        instrument_coverage_flags.rustflags.push(format!(
-            "profiler_builtins_no_core={}",
-            profiler_builtins_no_core_path.to_str().unwrap()
-        ));
-        instrument_coverage_flags
-            .rustflags
-            .push("-Zprofiler_runtime=profiler_builtins_no_core".into());
-        instrument_coverage_flags.rustflags.push("-L".into());
-        instrument_coverage_flags.rustflags.push(lib_dir.to_str().unwrap().to_string());
+        let mut cmd = BootstrapCommand::new(self.llvm_bin_dir.join("llvm-profdata"));
+        cmd.arg("merge").arg("--sparse").arg("-o").arg(self.profdata_file);
```

Comment:
I _think_ you can just supply the directory instead of the individual files.

```suggestion
        cmd.arg("merge").arg("--sparse").arg("-o").arg(self.profdata_file).arg(self.profraw_dir);
```

Reply:
Yes that works, didn't know about that! I'll update the code to reflect this.